WebExtension: Use API::Error instead of NSError

ItsJamie9494 opened the pull request:
#### e958140e35a329877004c3da7c86213083a8ce20
<pre>
WebExtension: Use API::Error instead of NSError
<a href="https://webkit.org/b/280948">https://webkit.org/b/280948</a>

Reviewed by Timothy Hatcher.

This replaces most uses of NSError with API::Error, along with updates localizations.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
(-[_WKWebExtensionLocalization _localizationDictionaryForWebExtension:withLocale:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
(+[WKWebExtension extensionWithAppExtensionBundle:completionHandler:]):
(+[WKWebExtension extensionWithResourceBaseURL:completionHandler:]):
(-[WKWebExtension _initWithAppExtensionBundle:resourceBaseURL:error:]):
(-[WKWebExtension errors]):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
(WebKit::WebExtensionContext::createInjectedContentForScripts):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::icon):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension):
(WebKit::WebExtension::parseManifest):
(WebKit::WebExtension::manifest):
(WebKit::WebExtension::resourceStringForPath):
(WebKit::WebExtension::resourceDataForPath):
(WebKit::WebExtension::recordError):
(WebKit::WebExtension::populateActionPropertiesIfNeeded):
(WebKit::WebExtension::imageForPath):
(WebKit::WebExtension::bestImageInIconsDictionary):
(WebKit::WebExtension::bestImageForIconsDictionaryManifestKey):
(WebKit::WebExtension::bestImageForIconVariants):
(WebKit::WebExtension::bestImageForIconVariantsManifestKey):
(WebKit::WebExtension::parseDeclarativeNetRequestRulesetDictionary):
(WebKit::WebExtension::populateDeclarativeNetRequestPropertiesIfNeeded):
(WebKit::toAPI): Deleted.
(WebKit::WebExtension::createError): Deleted.
(WebKit::WebExtension::errors): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::errors):
(WebKit::WebExtensionContext::addInjectedContent):
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRules):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
(WebKit::WebExtensionDynamicScripts::sourcePairForResource):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
(WebKit::WebExtensionMenuItem::icon const):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask):
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::toAPI):
(WebKit::WebExtension::createError):
(WebKit::WebExtension::errors):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::createError):
(WebKit::WebExtension::recordErrorIfNeeded):

Canonical link: <a href="https://commits.webkit.org/284783@main">https://commits.webkit.org/284783@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60e64c8c898114fb8403b0b13ec3f7f9cb9fc4a1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70516 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49922 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23281 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74605 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/21694 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72633 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57722 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21534 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/74605 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/21694 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73582 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/45405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60789 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74605 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/42064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK1-Tests-EWS "Waiting to run tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/20055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/63997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/18576 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76324 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/14742 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Debug-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63596 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/14785 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60856 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63530 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15607 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45724 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/46798 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/48075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46540 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -361,9 +363,9 @@
-                recordError(error);
+            RefPtr<API::Error> error;
+            if (!extension->resourceStringForPath(scriptPath, error, WebExtension::CacheResult::No, WebExtension::SuppressNotFoundErrors::Yes)) {
+                recordError(::WebKit::wrapper(*error));
```
Shouldn't need `::WebKit::` or `*`. It wasn't needed in the other cases in the PR.
```suggestion
                recordError(wrapper(error));
```

ItsJamie9494 replied:
Without ::WebKit::, the compiler uses WebExtensionContext::wrapper(), which is obviously wrong. This also goes for WebExtension.cpp/WebExtensionCocoa.mm, where WebExtension::wrapper() is called instead. The pointer can be removed though.

xeenon replied:
Bummer! Might be able to do `error->wrapper()` instead. But `::WebKit::wrapper()` is fine for now. Converting `WebExtensionContext` to also use `API::Error` will eliminate this eventually.

ItsJamie9494 replied:
Yes, a lot of these cases will be eliminated when more code is ported, so I'm not terribly concerned about it. I can try error->wrapper() if that's better though

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
```diff
@@ -376,9 +378,9 @@
-                recordError(error);
+            RefPtr<API::Error> error;
+            if (!extension->resourceStringForPath(styleSheetPath, error, WebExtension::CacheResult::No, WebExtension::SuppressNotFoundErrors::Yes)) {
+                recordError(::WebKit::wrapper(*error));
```
Ditto.
```suggestion
                recordError(wrapper(error));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
```diff
@@ -736,14 +737,14 @@ - (void)_otherPopoverWillShow:(NSNotification *)notification
-            result = extensionContext()->extension().bestImageForIconVariants(m_customIconVariants.get(), idealSize, [&](auto *error) {
-                extensionContext()->recordError(error);
+            result = extensionContext()->extension().bestImageForIconVariants(m_customIconVariants.get(), idealSize, [&](Ref<API::Error> error) {
+                extensionContext()->recordError(::WebKit::wrapper(error));
```
```suggestion
                extensionContext()->recordError(wrapper(error));
```

ItsJamie9494 replied:
As mentioned above, just using wrapper calls WebExtensionAction::wrapper instead of the wrapper we want

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
```diff
@@ -736,14 +737,14 @@ - (void)_otherPopoverWillShow:(NSNotification *)notification
-            result = extensionContext()->extension().bestImageInIconsDictionary(m_customIcons.get(), idealSize, [&](auto *error) {
-                extensionContext()->recordError(error);
+            result = extensionContext()->extension().bestImageInIconsDictionary(m_customIcons.get(), idealSize, [&](Ref<API::Error> error) {
+                extensionContext()->recordError(::WebKit::wrapper(error));
```
```suggestion
                extensionContext()->recordError(wrapper(error));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -140,13 +141,11 @@
-        if (outError)
-            *outError = m_errors.get().lastObject;
+        ASSERT(!m_errors.isEmpty());
+        outError = m_errors.takeLast();
```
`takeLast()` mutates the Vector and removes the last. We don't want that here.
```suggestion
        outError = m_errors.last();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -173,7 +172,8 @@
     if (!m_manifest) {
-        recordError(createError(Error::InvalidManifest, { }, parseError));
+        if (parseError)
+            recordError(createError(Error::InvalidManifest, { }, API::Error::create(parseError))); // Convert parseError to an API::Error
```
```suggestion
            recordError(createError(Error::InvalidManifest, { }, API::Error::create(parseError)));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -309,7 +310,10 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 NSArray *WebExtensionContext::errors()
 {
-    return [protectedExtension()->errors() arrayByAddingObjectsFromArray:m_errors.get()];
+    auto array = createNSArray(protectedExtension()->errors(), [](Ref<API::Error> error) {
+        return ::WebKit::wrapper(error);
+    });
```
```suggestion
    auto *array = createNSArray(protectedExtension()->errors(), [](Ref<API::Error> error) {
        return wrapper(error);
    });
```

ItsJamie9494 replied:
`error: variable 'array' with type 'auto *' has incompatible initializer of type 'RetainPtr<NSMutableArray>'`

xeenon replied:
Ah, always forget it returns `RetainPtr`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -329,7 +333,7 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
         RELEASE_LOG_ERROR(Extensions, "Cannot load persistent background content on this platform");
         if (outError)
-            *outError = extension->createError(WebExtension::Error::InvalidBackgroundPersistence);
+            *outError = ::WebKit::wrapper(extension->createError(WebExtension::Error::InvalidBackgroundPersistence)).get();
```
```suggestion
            *outError = wrapper(extension->createError(WebExtension::Error::InvalidBackgroundPersistence)).get();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -4347,10 +4351,10 @@ void effectiveAppearanceDidChange(Inspector::ExtensionAppearance appearance) ove
+            auto *scriptString = extension->resourceStringForPath(scriptPath, error, WebExtension::CacheResult::Yes);
             if (!scriptString) {
-                recordError(error);
+                recordError(::WebKit::wrapper(error));
```
```suggestion
                recordError(wrapper(error));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -4371,10 +4375,10 @@ void effectiveAppearanceDidChange(Inspector::ExtensionAppearance appearance) ove
+            auto *styleSheetString = extension->resourceStringForPath(styleSheetPath, error, WebExtension::CacheResult::Yes);
             if (!styleSheetString) {
-                recordError(error);
+                recordError(::WebKit::wrapper(error));
```
```suggestion
                recordError(wrapper(error));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -4640,10 +4644,10 @@ void effectiveAppearanceDidChange(Inspector::ExtensionAppearance appearance) ove
+            auto *jsonData = extension->resourceDataForPath(ruleset.jsonPath, error);
             if (!jsonData) {
-                recordError(error);
+                recordError(::WebKit::wrapper(*error));
```
```suggestion
                recordError(wrapper(*error));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
```diff
@@ -82,10 +84,10 @@ static bool userStyleSheetMatchesContent(Ref<API::UserStyleSheet> userStyleSheet
+    auto *scriptData = extensionContext.protectedExtension()->resourceDataForPath(path, error);
     if (!scriptData) {
-        extensionContext.recordError(error);
+        extensionContext.recordError(::WebKit::wrapper(error));
```
```suggestion
        extensionContext.recordError(wrapper(error));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
```diff
@@ -367,14 +369,14 @@ + (BOOL)usesUserKeyEquivalents
-        result = extensionContext()->protectedExtension()->bestImageForIconVariants(m_iconVariants.get(), idealSize, [&](auto *error) {
-            extensionContext()->recordError(error);
+        result = extensionContext()->protectedExtension()->bestImageForIconVariants(m_iconVariants.get(), idealSize, [&](Ref<API::Error> error) {
+            extensionContext()->recordError(::WebKit::wrapper(error.get()));
```
```suggestion
            extensionContext()->recordError(wrapper(error.get()));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
```diff
@@ -367,14 +369,14 @@ + (BOOL)usesUserKeyEquivalents
-        result = extensionContext()->protectedExtension()->bestImageInIconsDictionary(m_icons.get(), idealSize, [&](auto *error) {
-            extensionContext()->recordError(error);
+        result = extensionContext()->protectedExtension()->bestImageInIconsDictionary(m_icons.get(), idealSize, [&](Ref<API::Error> error) {
+            extensionContext()->recordError(::WebKit::wrapper(error.get()));
```
```suggestion
            extensionContext()->recordError(wrapper(error.get()));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
```diff
@@ -104,10 +106,10 @@
+        auto *fileData = extensionContext->extension().resourceDataForPath(requestURL.path().toString(), error);
         if (!fileData) {
-            extensionContext->recordError(error);
+            extensionContext->recordError(::WebKit::wrapper(*error));
```
```suggestion
            extensionContext->recordError(wrapper(*error));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -237,6 +245,189 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+        actionManifest = manifestObject()->getValue(browserActionManifestKey);
+        if (!actionManifest)
+            actionManifest = manifestObject()->getValue(pageActionManifestKey);
+    }
```
This should move since `case Error::InvalidActionIcon: { … }` since it is only needed there.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -237,6 +245,189 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+            localizedDescription = WEB_UI_STRING("Missing or empty `browser_action` or `page_action` manifest entry.", "WKWebExtensionErrorInvalidManifestEntry description for browser_action or page_action");
+        break;
+
+    case Error::InvalidActionIcon:
```
You can put it here with a brace, like this:
```suggestion
    case Error::InvalidActionIcon: {
        RefPtr<JSON::Value> actionManifest;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -237,6 +245,189 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
+            localizedDescription = WEB_UI_STRING("Empty or invalid `default_icon` for the `action` manifest entry.", "WKWebExtensionErrorInvalidManifestEntry description for default_icon in action only");
+        else
+            localizedDescription = WEB_UI_STRING("Empty or invalid `default_icon` for the `browser_action` or `page_action` manifest entry.", "WKWebExtensionErrorInvalidManifestEntry description for default_icon in browser_action or page_action");
+        break;
```
```suggestion
        break;
    }
```

ItsJamie9494 committed:
WebExtension: Use API::Error instead of NSError
https://webkit.org/b/280948

Reviewed by Timothy Hatcher.

This replaces most uses of NSError with API::Error, along with updates localizations.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/Shared/Extensions/_WKWebExtensionLocalization.mm:
(-[_WKWebExtensionLocalization _localizationDictionaryForWebExtension:withLocale:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
(+[WKWebExtension extensionWithAppExtensionBundle:completionHandler:]):
(+[WKWebExtension extensionWithResourceBaseURL:completionHandler:]):
(-[WKWebExtension _initWithAppExtensionBundle:resourceBaseURL:error:]):
(-[WKWebExtension errors]):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIScriptingCocoa.mm:
(WebKit::WebExtensionContext::createInjectedContentForScripts):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::icon):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension):
(WebKit::WebExtension::parseManifest):
(WebKit::WebExtension::manifest):
(WebKit::WebExtension::resourceStringForPath):
(WebKit::WebExtension::resourceDataForPath):
(WebKit::WebExtension::recordError):
(WebKit::WebExtension::populateActionPropertiesIfNeeded):
(WebKit::WebExtension::imageForPath):
(WebKit::WebExtension::bestImageInIconsDictionary):
(WebKit::WebExtension::bestImageForIconsDictionaryManifestKey):
(WebKit::WebExtension::bestImageForIconVariants):
(WebKit::WebExtension::bestImageForIconVariantsManifestKey):
(WebKit::WebExtension::parseDeclarativeNetRequestRulesetDictionary):
(WebKit::WebExtension::populateDeclarativeNetRequestPropertiesIfNeeded):
(WebKit::toAPI): Deleted.
(WebKit::WebExtension::createError): Deleted.
(WebKit::WebExtension::errors): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::errors):
(WebKit::WebExtensionContext::addInjectedContent):
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRules):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm:
(WebKit::WebExtensionDynamicScripts::sourcePairForResource):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
(WebKit::WebExtensionMenuItem::icon const):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask):
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::toAPI):
(WebKit::WebExtension::createError):
(WebKit::WebExtension::errors):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::createError):
(WebKit::WebExtension::recordErrorIfNeeded):

Canonical link: https://commits.webkit.org/284783@main
